src/backend/distributed/planner/query_pushdown_planning_utils.c

```diff
@@ -0,0 +1,1704 @@
+/*-------------------------------------------------------------------------
+ *
+ * query_pushdown_planning_utils.c
+ *
+ * Routines for creating pushdown plans for queries.
+ *
+ * Copyright (c) 2012-2016, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "distributed/citus_clauses.h"
+#include "distributed/citus_ruleutils.h"
+#include "distributed/deparse_shard_query.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/multi_logical_optimizer.h"
+#include "distributed/multi_logical_planner.h"
+#include "distributed/multi_router_planner.h"
+#include "distributed/pg_dist_partition.h"
+#include "distributed/query_pushdown_planning_utils.h"
+#include "distributed/relation_restriction_equivalence.h"
+#include "nodes/nodeFuncs.h"
+#include "nodes/pg_list.h"
+#include "optimizer/clauses.h"
+#include "optimizer/var.h"
+#include "parser/parsetree.h"
+
+/* Local functions forward declarations */
+static bool JoinTreeContainsSubqueryWalker(Node *joinTreeNode, void *context);
+static bool IsFunctionRTE(Node *node);
+static bool IsNodeQuery(Node *node);
+static bool WindowPartitionOnDistributionColumn(Query *query);
+static DeferredErrorMessage * DeferErrorIfFromClauseRecurs(Query *queryTree);
+static DeferredErrorMessage * DeferredErrorIfUnsupportedRecurringTuplesJoin(
+	PlannerRestrictionContext *plannerRestrictionContext);
+static DeferredErrorMessage * DeferErrorIfUnsupportedTableCombination(Query *queryTree);
+static bool ShouldRecurseForRecurringTuplesJoinChecks(RelOptInfo *relOptInfo);
+static bool RelationInfoContainsRecurringTuples(PlannerInfo *plannerInfo,
+												RelOptInfo *relationInfo,
+												RecurringTuplesType *recurType);
+static bool IsRecurringRTE(RangeTblEntry *rangeTableEntry,
+						   RecurringTuplesType *recurType);
+static bool IsRecurringRangeTable(List *rangeTable, RecurringTuplesType *recurType);
+
+
+
+static MultiNode * SubqueryPushdownMultiNodeTree(Query *queryTree);
+static void FlattenJoinVars(List *columnList, Query *queryTree);
+static void UpdateVarMappingsForExtendedOpNode(List *columnList,
+											   List *subqueryTargetEntryList);
+static MultiTable * MultiSubqueryPushdownTable(Query *subquery);
+static List * CreateSubqueryTargetEntryList(List *columnList);
+static Task * QueryPushdownTaskCreate(Query *originalQuery, int shardIndex,
```

Comment:
In general, we try to follow the rule: The function definition and declaration orders should be consistent. So, if `QueryPushdownTaskCreate ` is defined later in the code, we generally declare it later (compared to the other functions obviously such as `ErrorIfUnsupportedShardDistribution `).

We follow the same rule for header file declarations as well such as `SubqueryMultiNodeTree` vs `QueryPushdownSqlTaskList`.

Since we're doing a refactoring, I'd be in favor of sticking our conventions.